decisions/0003-package-for-shared-server-features.md

```diff
@@ -0,0 +1,113 @@
+# Package for shared server features
+
+Date: 2022-06-01
+
+Status: proposed
+
+## Context
+
+Currently, users need to install 4 Remix dependencies:
+
+- Dev (`@remix-run/dev`)
+- Server runtime (`@remix-run/{cloudflare,deno,node}`)
+- Adapter (`@remix-run/{arc,cloudflare-pages,cloudflare-workers,express,fly,netlify,vercel}`)
+- Renderer (`@remix-run/react`)
+
+User code currently looks like this:
+```tsx
+// server runtime
+import { json, createFileSessionStorage } from "@remix-run/node";
+// adapter
+import { createRequestHandler } from "@remix-run/express"
+// renderer
+import { useLoaderData } from "@remix-run/react";
+```
+
+### Goals
+
+- make DX and docs simpler: more intuitive, less magic, eliminate some user errors
+- reduce bugs from `remix` internal complexity
+
+### Non-goals
+
+- maintainability of `remix` repo
+- more portable `app/` directory
+- reduce or increase number of packages
+
+Some of these are nice-to-haves, but not the primary reasons for this proposal.
+
+### Issues
+
+1. Users can import features from their runtime that aren't supported by their adapter:
+
+```tsx
+import { createFileSessionStorage } from "@remix-run/node";
+// Implementation only depends on server runtime, but _availability_ depends on the adapter!
+// ❌ This will crash in netlify nor in vercel!
+```
+
+2. Docs, examples, and videos refer to `@remix-run/node` by default for shared functionality like `json`:
+
+```tsx
+import { json } from "@remix-run/node";
+// Need docs to tell users that they should import from their corresponding runtime package, 
+// and not from `@remix-run/server-runtime`.
+```
+
+3. Keeping shared exports consistent across runtime packages is [error-prone](https://github.com/remix-run/remix/pull/3215) and tedious.
+
+## Decision
+
+Users should import features from _adapter packages_, not from runtime packages. Each adapter can export only the features it supports. Shared features will be imported from a new package, `@remix-run/server`.
+
+This means there won't be any imports needed from the runtime packages (`@remix-run/{cloudflare,deno,node`}) and these will become an implementation detail for most users. Only developers that need to develop adapters will import from the runtime packages directly.
+
+### User-facing packages
+
+Users still need to import from 4 Remix packages:
+
+- Dev `@remix-run/dev` (unchanged)
+- Shared server features `@remix-run/server`
```

Comment:
We currently already have a draft version of [`@remix-run/router`](https://github.com/remix-run/react-router/tree/19d5fa6ab66300ef529dcf39f8089f3869d6e843/packages/router) (which will be official once `react-router` v6.4 is released), which exports a lot of `@remix-run/server-runtime` (including functions like `json` & `redirect`).

Do we want to keep the functionality in `@remix-run/router` or do we still want to extract it into `@remix-run/server`?

Reply:
I thing those functions only make sense on the server when using Remix, since you will only use them there. So it will make sense to export them from `@remix-run/server`

Reply:
@sergiodxa `loaders` could return a `Response`, so _at least_ the `json` helper could be interesting to use on the frontend when using `react-router` as well.

CC/ @brophdawg11 @ryanflorence

Reply:
When using React Router you can use `json` and `redirect` no the client, but when using Remix there's no reason, your loader and action functions only run server-side.

Why would you want to create a Response using those helpers client-side in Remix?

Reply:
@sergiodxa It seems a bit weird to me as well, but it's explained in the docs
https://beta.reactrouter.com/en/dev/route/loader#returning-responses

Reply:
Yeah I understand that, my point is that makes sense in React Router, but in Remix there’s no need for using those response helpers client side, only server side, so they should be exported from the proposed @remix-run/server package only, not client side

Reply:
I'm not too familiar with the server adapter landscape - but `json`/`redirect` will indeed be exported from `@remix-run/router`.  I'm unsure if we'll actually leverage that package to do SSR data fetching when we bring remix on top of RR 6.4 but at the very least we can re-export them through the remix server packages so we're not duping them

Reply:
Yea I think `json` and similar shared server features from the proposed `@remix-run/server` will need some SSR-specific implementation on the Remix side. So then having RR export a client-side/SPA version of `json` (and friends) is fine alongside Remix's SSR-specific implementations.

If we find that we can generalize the implementation so that it doesn't depend on whether that feature is used client- or server-side, then we can implement where ever  we think is best and re-export it as needed.

Reply:
But there's no server-side or client-side versions of `json` and `redirect`, they just create a `Response`, that could work server-side and client-side, and the `Response` is global on both the server (in Remix) and the browser

Reply:
Yea I'm just referring to `json` and `redirect` as examples of shared server features/exports. Those ones aren't server-side nor client-side specific, but I'm just saying that even if there _were_ features that needed to be client/server-side specific, we'd still have a way to deal with that too.

So just saying that we aren't painting ourselves into a corner.